### ISSUE TITLE
New Dashboard: Cannot login with google.

### ISSUE THREAD

**Issue by farmaazon** [enso-org/enso#6658]
### Discord username

farmaazon

### What type of issue is this?

Permanent – Occurring repeatably

### Is this issue blocking you from using Enso?

- [ ] Yes, I can't use Enso because of this issue.

### Is this a regression?

- [ ] Yes, previous version of Enso did not have this issue.

### What issue are you facing?

Logging with Google option does not work.

### Expected behaviour

Google is able to authenticate me to Enso.

### How we can reproduce it?

I click "Login with Google" button, what opens a tab in my browser (Opera) which in turn asks if it could run xdg-open application; Once I agree, nothing happens.

![image](https://github.com/enso-org/enso/assets/3919101/3b522548-c948-4456-bec7-d8bc1a0c4d11)

### Enso Version

9e71fea844920ecedcc93ca774e33ad40d4576a8

### Browser or standalone distribution

Standalone distribution

### Browser Version or standalone distribution

standalone

### Operating System

Linux

### Operating System Version

Garuda Linux 6.3.1-zen1-1-zen

**Comment by Akirathan** [comment]
On any Linux distribution, it is problematic to support `xdg-open` for enso, unless we provide some kind of installation that will update the `xdg` MIME type database. For example, on my distribution, Fedora 38, the logging in with Google looks like this:
[Screencast from 2023-05-15 13-38-18.webm](https://github.com/enso-org/enso/assets/14013887/bb9f7005-d1eb-4d63-b156-2512442c4b7b)

Which clearly shows that `xdg-open` fails for "Enso protocol"

**Comment by somebody1234** [comment]
Yeah - this is a known issue. It is easy enough to add a `.desktop` file that registers the protocol association though, however since we are building `AppImage`s, we cannot actually install a desktop file given that the executable doesn't necessarily have a consistent location.

See also: #6147, #5239

**Comment by JaroslavTulach** [comment]
The IDE is an Electron application. It has access to all node.js API including ability to start an HTTP server. Why don't we let the IDE start a server on `localhost:xyz` and then redirect the login result to that server? This wouldn't require any global registrations in the Linux system.

**Comment by somebody1234** [comment]
@JaroslavTulach that doesn't sound like a bad idea actually - especially since we already are running a server to host all the content. some minor decisions though:
- should this special handling be around, even for `./run ide build` (and on all platforms too)? it'd be simpler to do things that way, but not sure whether it's a good idea to always keep that functionality around.
- the server runs on the first found free port, rather than any fixed port. this is an issue, as the authentication requires each host (hostname + port) to be manually whitelisted. i've suggested some solutions in another issue - i'll repeat them here for convenience:
  - (suggested previously) go through `cloud.enso.org` by passing custom state specifying the *actual* redirect url
  - use a fixed port that's uncommonly used, similar to what the Project Manager currently does
  - whitelist a range of ports and just not support ports outside that range

**Comment by wdanilo** [comment]
@somebody1234 regarding ports - hardcoding a port is a bad idea and we should not allow it. We can whitelist a range of ports (like 1000 or 10000 of them ;) - but then we need to double check our port finding algorithm to be sure it is following the same port range ). What do you mean by `go through cloud.enso.org by passing custom state specifying the actual redirect url` - could you elaborate more on this idea, please? 

Regarding login the way Jaroslav described it, it should work, but FYI, Jaroslav, this has one downside in comparison to what we've got on macOS and Windows now - there we are registering the app during installation, so deep links work even if the app is closed.

**Comment by somebody1234** [comment]
going through cloud.enso.org:
`cloud.enso.org/?normal_oauth_callback_parameters=here&state=normal_state-custom_state`
where `custom_state` would encode the port to redirect to

before going ahead with whitelisting a ton of ports though, we should first consider whether this is a security risk - although it's not like it's not already one, because anyone could theoretically register the `enso` protocol, or start up an application on port 8080, and it's not like we aren't in control of which redirect we're pointing to

as for the macos and windows thing: that's true, but i'm think this is being pitched as a way for logins to work when deep links aren't an option (so linux and other unixes, and `./run ide watch`)

**Comment by somebody1234** [comment]
but yeah, considering [this](https://github.com/enso-org/enso/issues/6730#issuecomment-1551662603), hardcoding a port does seem like it may cause an issue if, say, multiple users are running enso at the same time

**Comment by somebody1234** [comment]
@wdanilo after looking at the [cognito docs](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_UpdateUserPoolClient.html#CognitoUserPools-UpdateUserPoolClient-request-CallbackURLs) it seems like there's a hard limit of 100 whitelisted redirect URLs - note that this includes
- the `enso://` url
- `localhost:<port>`
- `localhost:<port>/callback` - this is the hypothetical endpoint for the server run by electron, such that when it receives a request to this endpoint, it completes the authentication flow for the main window (for Linux, `ide watch` and maybe `gui watch` too)

**Comment by wdanilo** [comment]
@somebody1234 Why whitelisting many ports could be a security risk? Would you be so nice and give me an example of that? If I understand correctly, our app sends request to our login server and tells it to what port it should redirect to. This port is occupied by our application and until someone does not kill our app in the meantime and spawns their own app at exactly the same port, it should be safe right? I consider this situation very unlikely and if it happens, it means that someone has already their computer infected with some kind of a virus nevertheless. 

Anyway, white listening 30 - 50 ports should make the thing work IMO. But when you whitelist, please make sure that our port searching algorithm follows the same port pattern. It would be good if these are NOT consecutive ports, as `8080` is insanely popular and there is a lot of apps starting at it and searching up for a free one. Probably we should create a math equation that will generate 50 ports and we should stick with it, you know, something like `for (let i=0; i<50; i++) { console.log((i+1)*13 - i * 7 + (i*6) * 31 + 3) }` (just random thing out of my head - including 8080 ;) )

**Comment by somebody1234** [comment]
yeah, that's fair - i just figured it's safer to assume that it may be a security risk, rather than finding out too late if it actually was one

as for not being consecutive ports - that's a good idea.
not too sure about that algorithm tbh... mostly because the ports 1024-49151 are "registered ports" - so their intended use is to be individually registered with IANA first, and then used as fixed ports
(it also starts below 1024 which normally requires root access to start servers on!)

i think a simple RNG may be the least likely to cause collisions? e.g:

```js
let x = parseInt('enso', 36)
const ports = []
for (let i = 0; i < 29; i++) {
    ports.push(49152 | x & 16383)
    // xorshift32
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
}
ports
```

(speaking of ports, i guess it's worth considering whether we really want to use port 8080 at all - especially considering it is probably the most likely out of almost every port other than 80 and 443, to already be occupied)

**Comment by wdanilo** [comment]
Oh, forgot about the reserved ports. Yeah, any algorithm that will give us some different ports is good. I'd not skip 8080 only because of one reason (and I will start always with 8080) - everyone is so used to having this port as part of the development process, that if we change it, we will make every dev's life (in our team) worse.
